File: src/main/java/io/github/svegon/capi/event/render/CameraSubmersionTypeCallback.java
package io.github.svegon.capi.event.render;

import net.fabricmc.fabric.api.event.Event;
import net.fabricmc.fabric.api.event.EventFactory;
import net.minecraft.client.render.Camera;
import net.minecraft.client.render.CameraSubmersionType;
import org.spongepowered.asm.mixin.injection.callback.CallbackInfoReturnable;

public interface CameraSubmersionTypeCallback {
    Event<CameraSubmersionTypeCallback> EVENT = EventFactory.createArrayBacked(CameraSubmersionTypeCallback.class,
            (camera, callback) -> {}, listeners -> (camera, callback) -> {
        for (CameraSubmersionTypeCallback listener : listeners) {
            listener.getCameraSubmersionType(camera, callback);

            if (callback.isCancelled()) {
                return;
            }
        }
    });

    void getCameraSubmersionType(Camera camera, CallbackInfoReturnable<CameraSubmersionType> callback);
}

File: src/client/kotlin/io/github/svegon/mclientapi/client/event/render/CameraSubmersionTypeCallback.kt
package io.github.svegon.mclientapi.client.event.render

import net.fabricmc.fabric.api.event.Event
import net.fabricmc.fabric.api.event.EventFactory
import net.minecraft.block.enums.CameraSubmersionType
import net.minecraft.client.render.Camera
import org.spongepowered.asm.mixin.injection.callback.CallbackInfoReturnable
import java.util.function.Function

fun interface CameraSubmersionTypeCallback {
    fun getCameraSubmersionType(camera: Camera, callback: CallbackInfoReturnable<CameraSubmersionType>)

    companion object {
        @JvmField
        val EVENT: Event<CameraSubmersionTypeCallback> = EventFactory.createArrayBacked(
            CameraSubmersionTypeCallback::class.java,
            CameraSubmersionTypeCallback { camera: Camera, callback: CallbackInfoReturnable<CameraSubmersionType> -> }
        ) { listeners: Array<CameraSubmersionTypeCallback> ->
            CameraSubmersionTypeCallback { camera: Camera, callback: CallbackInfoReturnable<CameraSubmersionType> ->
                for (listener in listeners) {
                    listener.getCameraSubmersionType(camera, callback)

                    if (callback.isCancelled()) {
                        return@CameraSubmersionTypeCallback
                    }
                }
            }
        }
    }
}
